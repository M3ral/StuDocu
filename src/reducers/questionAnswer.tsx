import {createAsyncThunk, createSelector, createSlice, PayloadAction} from "@reduxjs/toolkit";
import {IRootState} from "../store";

export interface NewQuestionAndAnswer {
  question: string
  answer: string
}

export interface NewQuestionAndAnswerAction extends NewQuestionAndAnswer {
}

export interface removeOneAndAnswerAction {
  id: string
}

export interface UpdateQuestionAndAnswerAction extends QuestionAndAnswer {
}

export interface QuestionAndAnswer {
  id: string
  question: string
  answer: string
}

export interface QuestionAnswerSlice {
  data: QuestionAndAnswer[]
  status: string | null
}

const initialState: QuestionAnswerSlice = {
  data: [
    {
      id: '1',
      question: 'Question 3',
      answer: 'Answer',
    },
    {
      id: '2',
      question: 'Question 2',
      answer: 'Answer',
    },
    {
      id: '3',
      question: 'Question',
      answer: 'Answer',
    },
  ],
  status: null,
}

export const addNew = createAsyncThunk('questionAnswer/addNew', (newQaA: NewQuestionAndAnswer) => {
  return new Promise(resolve => setTimeout(resolve, 5000, newQaA)) as Promise<NewQuestionAndAnswerAction>
})

const questionAndAnswerSlice = createSlice({
  name: 'questionAnswer',
  initialState,
  reducers: {
    addQuestionAndAnswer(state, action: PayloadAction<NewQuestionAndAnswerAction>) {
      const {question, answer} = action.payload
      state.data = [
        ...state.data,
        {
          id: new Date().valueOf().toString(),
          question, answer
        }
      ]
    },
    removeAllQuestionAndAnswers(state) {
      state.data = []
    },
    removeOneQuestionAndAnswer(state, action: PayloadAction<removeOneAndAnswerAction>) {
      const {id} = action.payload
      state.data = state.data.filter(({id: _id}) => _id !== id)
    },
    updateQuestionAndAnswer(state, action: PayloadAction<UpdateQuestionAndAnswerAction>) {
      const {id, question, answer} = action.payload

      const index = state.data.findIndex(({id: _id}) => _id === id)
      state.data[index] =
        {
          id,
          question,
          answer
        }
    },
    sortQuestions(state) {
      state.data = state.data.sort((a, b) => a.question.localeCompare(b.question))
    }
  },
  extraReducers: (builder) => {
    builder.addCase(addNew.pending, (state, action) => {
      state.status = 'loading'
    })
    builder.addCase(addNew.fulfilled, (state, action) => {
      const {question, answer} = action.payload
      state.data = [
        ...state.data,
        {
          id: new Date().valueOf().toString(),
          question,
          answer
        }
      ]
      state.status = 'completed'
    })
  }
})

export const totalQaAItems = createSelector(
  (state: IRootState) => state.questionsAndAnswers.data,
  (items) => items.length)

export const isLoading = createSelector(
  (state: IRootState) => state.questionsAndAnswers.status,
  (status) => status === 'loading')

export const {
  addQuestionAndAnswer,
  removeAllQuestionAndAnswers,
  removeOneQuestionAndAnswer,
  updateQuestionAndAnswer,
  sortQuestions,
} = questionAndAnswerSlice.actions


export default questionAndAnswerSlice.reducer
