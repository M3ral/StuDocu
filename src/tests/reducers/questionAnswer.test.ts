import questionAnswer, {
  addNew,
  addQuestionAndAnswer,
  QuestionAnswerSlice,
  removeAllQuestionAndAnswers,
  removeOneQuestionAndAnswer,
  sortQuestions,
  updateQuestionAndAnswer
} from "../../reducers/questionAnswer";
import {configureStore} from "@reduxjs/toolkit";

describe('reducers/questionAnswer', function () {
  it('should return the initial state', function () {
    expect(questionAnswer(undefined, {type: undefined})).toMatchInlineSnapshot(`
Object {
  "data": Array [
    Object {
      "answer": "Answer",
      "id": "1",
      "question": "Question 3",
    },
    Object {
      "answer": "Answer",
      "id": "2",
      "question": "Question 2",
    },
    Object {
      "answer": "Answer",
      "id": "3",
      "question": "Question",
    },
  ],
  "status": null,
}
`)
  });
  it('should handle a new Q&A being added to an empty list', function () {
    const previousState: QuestionAnswerSlice = {data: [], status: null}
    const mockProp = {
      question: 'Is this a question?',
      answer: 'Yes'
    }
    expect(questionAnswer(previousState, addQuestionAndAnswer(mockProp))).toMatchSnapshot({
      data: [
        {
          id: expect.any(String),
          question: 'Is this a question?',
          answer: 'Yes'
        }
      ],
      status: null
    }, "add new question");
  })

  it('should handle all the Q&A being deleted', function () {
    const previousState: QuestionAnswerSlice = {
      data: [
        {
          id: '1',
          question: 'Question 3',
          answer: 'Answer',
        },
        {
          id: '2',
          question: 'Question 2',
          answer: 'Answer',
        },
        {
          id: '3',
          question: 'Question',
          answer: 'Answer',
        },
      ],
      status: null,
    }
    expect(questionAnswer(previousState, removeAllQuestionAndAnswers())).toMatchObject({data: [], status: null})
  });
  it('should handle one Q&A being deleted', function () {
    const previousState: QuestionAnswerSlice = {
      data: [
        {
          id: '1',
          question: 'Question 3',
          answer: 'Answer',
        },
        {
          id: '2',
          question: 'Question 2',
          answer: 'Answer',
        },
        {
          id: '3',
          question: 'Question',
          answer: 'Answer',
        },
      ],
      status: null,
    }
    const expectedState: QuestionAnswerSlice = {
      data: [
        {
          id: '2',
          question: 'Question 2',
          answer: 'Answer',
        },
        {
          id: '3',
          question: 'Question',
          answer: 'Answer',
        },
      ],
      status: null,
    }
    expect(questionAnswer(previousState, removeOneQuestionAndAnswer({id: '1'}))).toEqual(expectedState)
  });
  it('should handle one Q&A being updated', function () {
    const previousState: QuestionAnswerSlice = {
      data: [
        {
          id: '1',
          question: 'Question 3',
          answer: 'Answer',
        },
        {
          id: '2',
          question: 'Question 2',
          answer: 'Answer',
        },
        {
          id: '3',
          question: 'Question',
          answer: 'Answer',
        },
      ],
      status: null,
    }
    const updatedQA = {
      id: '2',
      question: 'Was this question updated?',
      answer: 'Yes',
    }
    const expectedState: QuestionAnswerSlice = {
      data: [
        {
          id: '1',
          question: 'Question 3',
          answer: 'Answer',
        },
        updatedQA,
        {
          id: '3',
          question: 'Question',
          answer: 'Answer',
        },
      ],
      status: null,
    }
    expect(questionAnswer(previousState, updateQuestionAndAnswer(updatedQA))).toEqual(expectedState)
  });
  it('should handle questions being sorted', function () {
    const previousState: QuestionAnswerSlice = {
      data: [
        {
          id: '1',
          question: 'Question 3',
          answer: 'Answer',
        },
        {
          id: '2',
          question: 'Question 2',
          answer: 'Answer',
        },
        {
          id: '3',
          question: 'Question',
          answer: 'Answer',
        },
      ],
      status: null,
    }
    const expectedState: QuestionAnswerSlice = {
      data: [
        {
          id: '3',
          question: 'Question',
          answer: 'Answer',
        },
        {
          id: '2',
          question: 'Question 2',
          answer: 'Answer',
        },
        {
          id: '1',
          question: 'Question 3',
          answer: 'Answer',
        },
      ],
      status: null,
    }
    expect(questionAnswer(previousState, sortQuestions())).toEqual(expectedState)
  });
  it('should handle a new Q&A being added asynchronously', async () => {
    const previousState: QuestionAnswerSlice = {
      data: [
        {
          id: '1',
          question: 'Question 3',
          answer: 'Answer',
        },
        {
          id: '2',
          question: 'Question 2',
          answer: 'Answer',
        },
        {
          id: '3',
          question: 'Question',
          answer: 'Answer',
        },
      ],
      status: null,
    }
    const store = configureStore({
      reducer: function (state = previousState, action) {
        switch (action.type) {
          case 'questionAnswer/addNew/fulfilled': {
            const {question, answer} = action.payload
            return {
              ...state,
              data: [
                ...state.data,
                {
                  id: '4',
                  question,
                  answer
                }
              ]
            }
          }
          default:
            return state;
        }
      },
    });
    const mockProps = {
      question: 'Is this a question?',
      answer: 'Yes'
    }
    await store.dispatch(addNew(mockProps));
    const state = store.getState();
    expect(state).toEqual({
      data: [
        ...previousState.data,
        {
          id: '4',
          question: 'Is this a question?',
          answer: 'Yes'
        }
      ],
      status: null
    });
  }, 10000)
});
