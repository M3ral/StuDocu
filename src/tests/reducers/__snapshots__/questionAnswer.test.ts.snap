// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/questionAnswer should handle a new Q&A being added to an empty list: add new question 1`] = `
Object {
  "data": Array [
    Object {
      "answer": "Yes",
      "id": Any<String>,
      "question": "Is this a question?",
    },
  ],
  "status": null,
}
`;
